lib/Runtime/Executor/ThreadPoolExecutor.cpp

```diff
@@ -80,13 +80,38 @@ ExecutionState::ExecutionState(RunIdentifierTy id, const DAGNode *root,
     nodeParentsDone_[node] = 0;
 
     // Make an (empty) input context for the node.
-    auto ctx = llvm::make_unique<ExecutionContext>();
+    auto nodeInputCtx = llvm::make_unique<ExecutionContext>();
+
     if (resultTraceContext) {
-      ctx->setTraceContext(llvm::make_unique<TraceContext>(
+      nodeInputCtx->setTraceContext(llvm::make_unique<TraceContext>(
           resultTraceContext->getTraceLevel(),
           resultTraceContext->getTraceThread()));
     }
-    inputCtxs_.insert(std::make_pair(node, std::move(ctx)));
+
+    auto nodeInputPhBindings = nodeInputCtx->getPlaceholderBindings();
+
+    // Get the symbol table for the node.
+    const SymbolTableTy &symbolTable = node->runtimeBundle->getSymbolTable();
+
+    // Create Placeholders for the outputs of all intermediate nodes. These are
+    // not in the ExecutionContext passed to Executor::run, so they must be
+    // created by the Executor. Only output symbols need to be considered
+    // because the Placeholders created for them will be copied to the input
+    // PlaceholderBindings of child nodes by
+    // ThreadPoolExecutor::propagatePlaceholdersForNode.
+    for (const auto &symbolPair : symbolTable) {
+      const auto &symbolName = symbolPair.first;
+      const auto &symbolInfo = symbolPair.second;
+
+      if (symbolInfo.output) {
```

Comment:
The symbolTable contains more than placeholders. It is not enough to only check the "output" value so far.  A lot of non-phs will be added into `nodeInputPhBindings` now.

Reply:
Oh, I didn't know we have `symbolCategory` now.

Is it sufficient to check that `symbolCategory == SymbolCategory::Placeholder`?

Reply:
Unfortunately I don't think so, TensorViews take on the category of the tensor they are viewing, so a Placeholder tensorView would have category placeholder.

Reply:
@gcatron 

> Unfortunately I don't think so, TensorViews take on the category of the tensor they are viewing, so a Placeholder tensorView would have category placeholder.

So, do you say we have here a use-case where we need to know if a placeholder is a genuine placeholder or e.g. a ternsorview? If so, we could probably have separate enum cases like PlaceholderTensorView, etc. Or have a special boolean flag to indicate that a symbol is a placeholder.

Reply:
I think I like the PlaceholderTenserView and ConstantTensorView additions, I'll code it up. Then checking Placeholder will be fine.

Reply:
Once #2683 lands you'll be good to go.